### DISCUSSION TITLE
Throttle or buffer websocket messages from a subscription?

### DISCUSSION THREAD

**Discussion by eschaefer** [urql-graphql/urql#1370] [Q&A] [answered]
Hello all, I'm wondering about the best way to handle a flood of sudden messages in an `urql` subscription. For example, our service can sometimes dump a large number of websocket messages into a short amount of time...

<img width="475" alt="Screen Shot 2021-02-08 at 10 45 31" src="https://user-images.githubusercontent.com/547148/107230052-da5ca580-6a1e-11eb-8f63-1bdbdba2483d.png">

This ends up rendering the app kinda useless as 177 subscription updates happen in 7 seconds, basically triggering a ton of re-renders and locking things up.

Aside from making changes to the websocket service on the API, what could be done in `urql` to throttle or batch these updates every second or few seconds? I noticed wonka provides a [buffer](https://wonka.kitten.sh/api/operators#buffer) operator, but I'm not sure where to place it with relation to the subscription exchange.

**Comment by kitten** [comment, marked as answer]
I'll assume for the following that you're using our normalised cache, Graphcache. In general, these updates are really cheap there but indeed the re-renders are what can really add up as each update may trigger a flurry of more re-renders.

Generally this is why I'd say that strategic placements of `React.memo` and other optimisations in your tree are what you could look at first. If nothing has changed then nothing should re-render. Basically, in terms of re-renders, given that I don't know much about your app it will depend on your queries. How many queries update? What components do these queries update? Which elements then take a long time to re-render?

Regarding batching and throttling, this is a difficult topic. Batching becomes rather pointless at times when we're talking about changes that may impact many queries in different ways; in other words, we don't know whether it may be safe to drop or merge messages here. If it is then a custom exchange or custom logic around this can help a lot.
In such cases it may be effective for an exchange to be created that has a short time window of buffered messages that it combines (or where it drops some messages by throttling them).

This could be used to implement batched updates: https://wonka.kitten.sh/api/operators#buffer

Lastly, if all else fails, we could consider changes on our end to throttle subscription updates in Graphcache itself, which may make sense if we're considering the worst case scenario: that none of these messages can be dropped or merged and each will update a large query hundreds of times a second. This wouldn't cause a huge issue per-se but the only way to safely merge all of the "updates" on React's end is to delay the re-execution of queries.

There are some other ways to tackle this however:

1. You can debounce hook updates at the `useQuery` site. This may be complicated however as you wouldn't want to delay the first result.
2. It may be possible to replace the `dedup` exchange with one that debounces query updates when the source is already active; what I mean by that is that it may be possible to debounce outgoing queries that are already active, i.e. where the `useQuery` hook has already been mounted.

**Reply by eschaefer** [reply]
> I'll assume for the following that you're using our normalised cache, Graphcache [...]

Yes, indeed we are using Graphcache and are updating the cache via subscription updates in the cache exchange. This might be triggering the re-render cascade.

> [...] I'd say that strategic placements of React.memo and other optimisations in your tree are what you could look at first.

This sounds like the simplest solution, so I will look again at what can be done here! 

Since none of the messages can be dropped, I will rule out throttling updates in a custom exchange. However, I can try using the [buffer](https://wonka.kitten.sh/api/operators#buffer) operator to work on intervals in a custom exchange. 

Thank you so much for the thoughtful response. I can post a follow-up here later after I've tried a few of these options.

**Reply by kitten** [reply]
On our end, we've discussed whether we could throttle reexecutions in Graphcache itself. What worries me about that is that it adds more complexity to an already complex part of its codebase. But it's an option. On the other hand it may be simpler to focus on where the problem becomes apparent: React's render process. Concurrent is still a long ways off, but it may turn out that the excessive updates can be stopped in React itself before they clog up the main thread.
